### PR TITLE
Clarify some points in config lookup documentation

### PR DESCRIPTION
**Pull request by bcoca**
##### ISSUE TYPE

- Docs Pull Request

### CODE REVIEW

File: lib/ansible/plugins/lookup/config.py [hunk 1]
```diff
@@ -6,36 +6,41 @@
...
         type: string
         version_added: '2.12'
       plugin_name:
-        description: name of the plugin for which you want to retrieve configuration settings.
+        description: The name of the plugin for which you want to retrieve configuration settings.
         type: string
         version_added: '2.12'
       show_origin:
-        description: toggle the display of what configuration subsystem the value came from
+        description: This toggles controls if wer return or not what configuration subsystem the value came from
```
Comment (briantist) [inline lib/ansible/plugins/lookup/config.py]:
```suggestion
        description: Set this to return what configuration subsystem the value came from
```

File: lib/ansible/plugins/lookup/config.py [hunk 2]
```diff
@@ -6,36 +6,41 @@
...
+                     (defaults, config file, environment, CLI, or variables).
         type: bool
         version_added: '2.16'
+    notes:
+      - Be aware that currently this lookup cannot take keywords nor delegation into account,
+        so for options that support keywords or are affected by delegation, it is at best a good guess or aproximation.
```
Comment (briantist) [inline lib/ansible/plugins/lookup/config.py]:
```suggestion
        so for options that support keywords or are affected by delegation, it is at best a good guess or approximation.
```

**Comment by bcoca** [comment]
@felixfontein  i think i'm missing something in validate-modules change, anything obvious?

**Comment by felixfontein** [comment]
@bcoca the schema itself requires `choices` to be a list or a tuple: https://github.com/ansible/ansible/blob/8ce90b01deddac2983ca2008cd29dbeb9963e6c1/test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py#L299. You also need to allow `dict[Any, Any[str, [str]]]`; I don't remember the syntax for that out of my head though.

antsibull-docs should be able to handle this (https://github.com/ansible-community/antsibull-docs/blob/main/src/antsibull_docs/data/docsite/ansible-docsite/macros/choiceslist.rst.j2#L8-L20), but I'm not sure that other docs renderers (galaxy_ng, collection_prep, ansible-doc-extractor) support this syntax. While collection_prep seems to be abandoned and it's not clear to me whether ansible-doc-extractor is still actively used (just checked, the code and the default templates in https://github.com/felixfontein/ansible-doc-extractor/tree/master/src/ansible_doc_extractor do not seem to handle `choices` at all), galaxy_ng support is pretty important - and it doesn't support it right now: https://github.com/ansible/ansible-hub-ui/blob/master/src/components/render-plugin-doc/render-plugin-doc.tsx#L647-L689 only handles lists, not dicts. The first part of the rendering code handles `choices` not being a list (by not rendering anything), but the second part calls `choices.includes(defaul)` if `defaul` is truthy (which is problematic on its own since it ignores a default of `0` or an empty string, for example), which results in an `Uncaught TypeError: choices.includes is not a function` exception. I know too little about React so no idea whether it only breaks rendering of the default or even the whole docs for the plugins. In any case, this should get fixed / improved.

CC @himdel

**Comment by bcoca** [comment]
@felixfontein thanks, I'll create PRs for extractor and c_prep, for galaxy I'm relying on the galaxy team.

My main concern was buliding the doc site, but I know we already crossed that bridge before. I'm was not to worried for the others as I don't expect most collection authors to start using this format anytime soon, but I'll make sure to propagate the change soon.

File: test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py [hunk 3]
```diff
@@ -296,7 +296,7 @@ def argument_spec_schema(for_collection):
                 (is_callable, list_string_types),
                 [is_callable, list_string_types],
             ),
-            'choices': Any([object], (object,)),
+            'choices': Any([object], (object,), {object: Any[str,[str]]}),
```
Comment (felixfontein) [inline test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py]: Ah sorry, I think I pointed you to the wrong `choices` :facepalm:. This one is the argspec, not the documentation.

The right place is here: https://github.com/ansible/ansible/blob/689c905c6e5dfbc67817ff779c02cce24be14f27/test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py
This should probably be `'choices': Any([json_value], {json_value: doc_string_or_strings]}),`

File: test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py [hunk 4]
```diff
@@ -487,10 +487,17 @@ def check_option_choices(v):
...
         return v
 
-    for value in v_choices:
+    if isinstance(v_choices, dict):
+        # choices are still a list (the keys) but dict form serves to document each choice.
+        iterate = v_choices.keys()
+    else:
+        iterate = v_choices
+
+    for value in iterate:
```
Comment (felixfontein) [inline test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py:500]: Is this change necessary? `for value in v_choices` is equivalent to `for value in v_choices.keys()` if `v_choices` is a dict.

Comment (bcoca) [reply test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py:500]: python versions don't always agree on that, i played it safe since i didn't see a min python version

File: test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py [hunk 5]
```diff
@@ -542,7 +549,7 @@ def list_dict_option_schema(for_collection, plugin_type):
     basic_option_schema = {
         Required('description'): doc_string_or_strings,
         'required': bool,
-        'choices': list,
+        'choices': Any(list, dict),
```
Comment (felixfontein) [inline test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py]: I would really use
```suggestion
        'choices': Any(list, {json_value: doc_string_or_strings}),
```
here, otherwise there's nothing that makes sure that the values of the dict are strings or lists of strings.

Comment (bcoca) [reply test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py]: hmm, this actually breaks validation as 'unhashable'

Comment (felixfontein) [reply test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py]: How about using `object` instead of `json_value`? Or is that also not hashable?

Comment (bcoca) [reply test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py]: did just that, now on new error

File: test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py (hunk 4 above)
Comment (bcoca) [reply test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py:500]: i think i got it!

File: test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py [hunk 6]
```diff
@@ -542,7 +549,7 @@ def list_dict_option_schema(for_collection, plugin_type):
     basic_option_schema = {
         Required('description'): doc_string_or_strings,
         'required': bool,
-        'choices': list,
+        'choices': Any(list, {any_string_types: doc_string_or_strings}),
```
Comment (felixfontein) [inline test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py]:
```suggestion
        'choices': Any(list, {object: doc_string_or_strings}),
```
That way you can use dicts for choices also for other key types than strings.

Comment (felixfontein) [reply test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py]: (I thought I submitted this comment already during the weekend, but apparently I did not...)

Comment (bcoca) [reply test/lib/ansible_test/_util/controller/sanity/validate-modules/validate_modules/schema.py]: i swear i tried this iteration before ... but .. .passes now.. not looking back

**Comment by felixfontein** [comment]
Here is how it looks like on the docsite: https://docs.ansible.com/ansible/devel/collections/ansible/builtin/config_lookup.html#parameter-on_missing

If it would be a list of strings for a choice value, there would be multiple paragraphs inside the corresponding `<li>`.
